src/TSTransformer/nodes/expressions/transformYieldExpression.ts

```diff
@@ -12,14 +12,30 @@ export function transformYieldExpression(state: TransformState, node: ts.YieldEx
 	const expression = transformExpression(state, node.expression);
 	if (node.asteriskToken) {
 		const loopId = luau.tempId("result");
+		const returnValue = luau.tempId("return");
+
+		state.prereq(
+			luau.create(luau.SyntaxKind.VariableDeclaration, {
+				left: returnValue,
+				right: undefined,
+			}),
+		);
+
 		state.prereq(
 			luau.create(luau.SyntaxKind.ForStatement, {
 				ids: luau.list.make(loopId),
 				expression: luau.property(convertToIndexableExpression(expression), "next"),
 				statements: luau.list.make<luau.Statement>(
 					luau.create(luau.SyntaxKind.IfStatement, {
 						condition: luau.property(loopId, "done"),
-						statements: luau.list.make(luau.create(luau.SyntaxKind.BreakStatement, {})),
+						statements: luau.list.make<luau.Statement>(
+							luau.create(luau.SyntaxKind.Assignment, {
+								left: returnValue,
+								operator: "=",
+								right: luau.property(loopId, "value"),
+							}),
```

Comment:
When the final value isn't used, this implementation creates an unnecessary variable (and assignment to it). We can prevent this by checking `isUsedAsStatement(node)`.